base_layer/core/src/base_node/states/horizon_state_sync/state_sync.rs

```diff
@@ -0,0 +1,705 @@
+        for block_nums in block_height_range.chunks(config.header_request_size) {
+            for attempt in 1..=config.max_sync_request_retry_attempts {
+                let (headers, sync_peer) = helpers::request_headers(
+                    LOG_TARGET,
+                    self.shared,
+                    self.sync_peers,
+                    block_nums,
+                    self.shared.config.horizon_sync_config.max_header_request_retry_attempts,
+                )
+                .await?;
+
+                match self.validate_and_insert_headers(block_nums, headers).await {
+                    Ok(_) => {
+                        trace!(
+                            target: LOG_TARGET,
+                            "Headers successfully added to database: {:?}",
+                            block_nums
+                        );
+                        break;
+                    },
+                    Err(err @ HorizonSyncError::EmptyResponse) |
+                    Err(err @ HorizonSyncError::IncorrectResponse) |
+                    Err(err @ HorizonSyncError::InvalidHeader(_)) => {
+                        warn!(target: LOG_TARGET, "Peer `{}`: {}", sync_peer, err);
+                        debug!(
+                            target: LOG_TARGET,
+                            "Banning peer {} from local node, because they supplied an invalid response", sync_peer
+                        );
+                        self.ban_sync_peer(sync_peer).await?;
+                    },
+                    // Fatal
+                    Err(e) => return Err(e),
+                }
+
+                if attempt == config.max_sync_request_retry_attempts {
+                    debug!(target: LOG_TARGET, "Reached maximum ({}) attempts", attempt);
+                    return Err(HorizonSyncError::MaxSyncAttemptsReached);
+                }
+                debug!(
+                    target: LOG_TARGET,
+                    "Retrying header sync. Attempt {} of {}", attempt, config.max_sync_request_retry_attempts
+                );
+            }
+        }
+
+        Ok(())
+    }
+
+    // Synchronize kernels upto the horizon sync height from remote sync peers.
+    async fn synchronize_kernels(&mut self) -> Result<(), HorizonSyncError> {
+        let config = self.shared.config.horizon_sync_config;
+        let local_num_kernels =
+            async_db::fetch_mmr_node_count(self.db(), MmrTree::Kernel, self.horizon_sync_height).await?;
+        let (remote_num_kernels, sync_peer) = helpers::request_mmr_node_count(
+            LOG_TARGET,
+            self.shared,
+            self.sync_peers,
+            MmrTree::Kernel,
+            self.horizon_sync_height,
+            config.max_mmr_node_request_retry_attempts,
+        )
+        .await?;
+
+        if local_num_kernels >= remote_num_kernels {
+            debug!(target: LOG_TARGET, "Local kernel set already synchronized");
+            return Ok(());
+        }
+
```

Comment:
You can check the MMR roots of the kernel here.  That should be a quick easy comparison of two hashes, before we start downloading the kernels

Reply:
Good idea - there are a few iterations needed on this code, think this should included in a subsequent PR